consensus/istanbul/backend/internal/enodes/val_enode_db.go

```
@@ -212,78 +213,122 @@ func (vet *ValidatorEnodeDB) String() string {
+func (vet *ValidatorEnodeDB) GetAddressFromNodeID(nodeID enode.ID) (common.Address, error) {
+	vet.lock.RLock()
+	defer vet.lock.RUnlock()
+	return vet.getAddressFromNodeID(nodeID)
+}
+
+func (vet *ValidatorEnodeDB) getAddressFromNodeID(nodeID enode.ID) (common.Address, error) {
+	rawEntry, err := vet.db.Get(nodeIDKey(nodeID), nil)
 	if err != nil {
 		return common.ZeroAddress, err
 	}
 	return common.BytesToAddress(rawEntry), nil
 }
 
+func (vet *ValidatorEnodeDB) GetAllValEnodes() (map[common.Address]*AddressEntry, error) {
+	vet.lock.RLock()
+	defer vet.lock.RUnlock()
+	var entries = make(map[common.Address]*AddressEntry)
+
+	err := vet.iterateOverAddressEntries(func(address common.Address, entry *AddressEntry) error {
+		entries[address] = entry
+		return nil
+	})
+
+	if err != nil {
+		vet.logger.Error("ValidatorEnodeDB.GetAllAddressEntries error", "err", err)
+		return nil, err
+	}
+
+	return entries, nil
+}
+
 // Upsert will update or insert a validator enode entry; given that the existing entry
 // is older (determined by view parameter) that the new one
-func (vet *ValidatorEnodeDB) Upsert(remoteAddress common.Address, enodeURL string, view *istanbul.View) error {
+func (vet *ValidatorEnodeDB) Upsert(valEnodeEntries map[common.Address]*AddressEntry) error {
 	vet.lock.Lock()
 	defer vet.lock.Unlock()
 
-	currentEntry, err := vet.getAddressEntry(remoteAddress)
-	isNew := err == leveldb.ErrNotFound
+	batch := new(leveldb.Batch)
+	peersToRemove := make([]*enode.Node, 0, len(valEnodeEntries))
+	peersToAdd := make(map[common.Address]*enode.Node)
 
-	// Check errors
-	if !isNew && err != nil {
-		return err
-	}
+	for remoteAddress, addressEntry := range valEnodeEntries {
+		currentEntry, err := vet.getAddressEntry(remoteAddress)
+		isNew := err == leveldb.ErrNotFound
 
-	// If it is an old message, ignore it.
-	if err == nil && view.Cmp(currentEntry.view) <= 0 {
-		return errOldAnnounceMessage
-	}
+		// Check errors
+		if !isNew && err != nil {
+			return err
```

Should we return an error? Or continue?

---

The thought is that `Upsert` will behave like a transaction.  All entries must be successful, or none will be.